federation/pkg/federation-controller/sync/controller.go

```diff
@@ -336,24 +340,54 @@ func (s *FederationSyncController) reconcile(namespacedName types.NamespacedName
 	if err != nil {
 		glog.Errorf("Failed to get cluster list: %v", err)
 		s.deliver(namespacedName, s.clusterAvailableDelay, false)
-		return
+		return err, reconcileResult
 	}
 
 	operations := make([]util.FederatedOperation, 0)
 	for _, cluster := range clusters {
+		glog.V(3).Infof("Processing cluster: %s for %s: %s", cluster.Name, kind, key)
+
 		clusterObj, found, err := s.informer.GetTargetStore().GetByKey(cluster.Name, key)
 		if err != nil {
 			glog.Errorf("Failed to get %s from %s: %v", key, cluster.Name, err)
 			s.deliver(namespacedName, 0, true)
-			return
+			return err, reconcileResult
 		}
+		reconcileResult[cluster.Name] = found
 
 		// The data should not be modified.
 		desiredObj := s.adapter.Copy(obj)
+		objMeta := s.adapter.ObjectMeta(obj)
+
+		// Check to see if this should be sent to the cluster.
+		send, err := util.SendToCluster(cluster.Labels, objMeta.Annotations)
+		if !send && err == nil {
+			glog.V(5).Infof("Skipping cluster: %s for %s: %s reason: cluster selectors do not match: %-v %-v", cluster.Name, kind, key, cluster.ObjectMeta.Labels, objMeta.Annotations[federationapi.FederationClusterSelector])
+			reconcileResult[cluster.Name] = send
+
+			// Only delete objects from target clusters if they exist
+			if found {
+				s.eventRecorder.Eventf(obj, api.EventTypeNormal, "RemoveFromCluster", "Removing %s %s from cluster %s", cluster.Name, namespacedName, kind)
```

Comment:
RemoveFromCluster -> DeleteInCluster to be consistent with CreateInCluster and UpdateInCluster?